fontforge/splineutil.c

```diff
@@ -1110,81 +1110,118 @@ void SplinePointCategorize(SplinePoint *sp) {
 	else if (( nclen==0 && pclen!=0 && (cross = pcdir.x*ndir.y-pcdir.y*ndir.x)<bounds && cross>-bounds ) ||
 		( pclen==0 && nclen!=0 && (cross = ncdir.x*pdir.y-ncdir.y*pdir.x)<bounds && cross>-bounds ))
-	    sp->pointtype = pt_tangent;
+	    pt = pt_tangent;
 
-	/* If a point started out hv, and could still be hv, them make it so */
-	/*  but don't make hv points de novo, Alexey doesn't like change */
-	/*  (this only works because hv isn't a default setting, so if it's */
-	/*   there it was done intentionally) */
-	if ( sp->pointtype == pt_curve && oldpointtype == pt_hvcurve &&
+	if (pt == pt_curve &&
 		((sp->nextcp.x==sp->me.x && sp->prevcp.x==sp->me.x && sp->nextcp.y!=sp->me.y) ||
 		 (sp->nextcp.y==sp->me.y && sp->prevcp.y==sp->me.y && sp->nextcp.x!=sp->me.x)))
-	    sp->pointtype = pt_hvcurve;
+	    pt = pt_hvcurve;
     }
+    return pt;
 }
 
 int SplinePointIsACorner(SplinePoint *sp) {
-    enum pointtype old = sp->pointtype, new;
+	return SplinePointCategory(sp) == pt_corner;
+}
+
+enum pointtype SplinePointDowngrade(int current, int geom) {
+	enum pointtype np = current;
 
-    SplinePointCategorize(sp);
-    new = sp->pointtype;
-    sp->pointtype = old;
-return( new==pt_corner );
+	if ( current==pt_curve && geom!=pt_curve ) {
+		if ( geom==pt_hvcurve )
+			np = pt_curve;
+		else 
+			np = pt_corner;
+	} else if ( current==pt_hvcurve && geom!=pt_hvcurve ) {
+		if ( geom==pt_curve )
+			np = pt_curve;
+		else
+			np = pt_corner;
+	} else if ( current==pt_tangent && geom!=pt_tangent ) {
+		np = pt_corner;
+	}
+
+	return np;
 }
 
-void SPLCategorizePoints(SplinePointList *spl) {
-    Spline *spline, *first, *last=NULL;
+// Assumes flag combinations are already verified. Only returns false
+// when called with check_compat
+int _SplinePointCategorize(SplinePoint *sp, int flags) {
+	enum pointtype geom, dg, cur;
 
-    for ( ; spl!=NULL; spl = spl->next ) {
-	first = NULL;
-	for ( spline = spl->first->next; spline!=NULL && spline!=first; spline=spline->to->next ) {
-	    SplinePointCategorize(spline->from);
-	    last = spline;
-	    if ( first==NULL ) first = spline;
+	if ( flags & pconvert_flag_none )
+		// No points selected for conversion -- keep type as is
+		return true;
+	if ( flags & pconvert_flag_smooth && sp->pointtype == pt_corner )
+		// Convert only "smooth" points, not corners
+		return true;
+
+	geom = SplinePointCategory(sp);
+	dg = SplinePointDowngrade(sp->pointtype, geom);
+
+	if ( flags & pconvert_flag_incompat && sp->pointtype == dg )
+		// Only convert points incompatible with current type
+		return true;
+
+	if ( flags & pconvert_flag_by_geom ) {
+		if ( ! ( flags & pconvert_flag_hvcurve ) && geom == pt_hvcurve )
+			sp->pointtype = pt_curve;
+		else
+			sp->pointtype = geom;
+	} else if ( flags & pconvert_flag_downgrade ) {
+		sp->pointtype = dg;
+	} else if ( flags & pconvert_flag_force_type ) {
+		if ( sp->pointtype != dg ) {
+			cur = sp->pointtype;
+			sp->pointtype = dg;
+			SPChangePointType(sp,cur);
+		}
```

Comment:
No one is likely to want to do this without better point conversion code, but FontForge could eventually have better point conversion code ...